Mean-field TEMPO with multiple types of system

piperfw opened the pull request:
Functionality to allow for multiple systems with time and field dependent Hamiltonians to couple to a common field in a combined 'MeanFieldSystem.' Original design and implementation for PT-TEMPO by Joel Beckles (@JoelANB), later TEMPO backend and tidying by Piper Fowler-Wright (@piperfw),

_Note this contribution changes the way in which the mean-field functionality is invoked and so will break user scripts based on previous code._ These changes are reflected in the updated tutorial notebook and documentation.

gefux reviewed (commented):
Well done!

Beside the code comments below, here are some more:

1. I've made a few minor typo corrections I spotted and I've merged the `BaseTempo` and `Tempo` class in `tempo.py`. I couldn't push it to JoelANB's repository, so I've pushed it to my repo https://github.com/gefux/OQuPy/tree/pr-mean-field
2. The API overview in `/docs/pages/api.rst` needs to be updated.
3. The commit messages don't follow the standard format. They should start with a title of max 50 characters and then (after a blank line) an optional description or more detailed list of changes. I'd use `git rebase -i [earliest commit hash]` and then "reword" the commit messages to follow that format.
4. Given that this PR will make breaking changes that will need a new version, I'd either make the necessary changes on the pull request, or merge it into a development branch `dev-0.4.0`. Merging breaking changes into the main branch when they are not ready for publishing would make publishing bug fix versions (such as 0.3.4, 0.3.5) quite unpleasant.

piperfw commented on examples/mean-field_multiple_system.py:
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python
```
I'll change this

piperfw commented on examples/single-system.py:
```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python
+
```
In fact, single-system.py and test.py were old testing scripts that should have been removed at a previous stage (I suspect I failed to git rm), the three 'mean-field...py' examples were the versions I intended to keep. Thanks for spotting this.

piperfw replied:
Similarly 'tempo.py' shouldn't be present. This is a little concerning, because these are quite old example scripts and I remember tidying/removing the examples directory!

piperfw commented on examples/test.py:
```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python
```
See previous comment

piperfw commented on oqupy/contractions.py:
```diff
@@ -143,60 +232,95 @@ def compute_dynamics_with_field(
+        #  perform first order Runge-Kutta calculation
+        rk1 = mean_field_system.field_eom(t, state_list, field)
+        if next_state_list is None:
+            return rk1 * dt
```
Yes, this wasn't picked up in testing because as you note below the 'rk1' version is actually never used

piperfw commented on oqupy/contractions.py:
```diff
@@ -143,60 +232,95 @@ def compute_dynamics_with_field(
-                return rk1 * dt
-            rk2 = system.field_eom(t + dt, next_state, field + rk1 * dt)
-            return field + dt * (rk1 + rk2) / 2
+    def compute_field(t: float, dt: float, state_list: List[ndarray],
```
I agree and am happy with the simplified function (in two places)

piperfw replied:
There is a difference between `compute_field` (`contractions.py`) and `_compute_field` (`tempo.py`): in the latter the states (or at least those in `next_state_list`) need to be reshaped as the tempo backend returns a one-dimensional array but the system `field_eom` expects a square matrix.

The behaviour between tempo and PT tempo seems a bit inconsistent here, we may want to consider revising one or the other (i.e. always reshaping the state to a square matrix, or not).

piperfw commented on oqupy/contractions.py:
```diff
@@ -207,91 +331,143 @@ def controls(step: int):
+                times=list(times), system_states_list=system_states_list,
+                fields=field_list)
 
 def _get_propagators(
```
This is a nice reduction. I have created public `get_propagators` methods for each system type to be called in `contractions.py` or `tempo.py` where appropriate. See also comment RE issue #41

gefux replied:
That sounds like just the right solution. Thank you!

piperfw commented on oqupy/system.py:
```diff
@@ -386,16 +383,57 @@ def gammas(self) -> List[Callable[[float], float]]:
+
+        super().__init__(name, description)
+        tmp_system_list = _check_mean_field_system_list(system_list)
+        self.system_list = tmp_system_list
```
I wasn't sure whether the property definition should return a copy. I don't think it should, since it's a public attribute, compared to the `_field_eom` which is private (and so copied when accessed as `field_eom`). Useful to get some experience of when to use the Python descriptors

piperfw commented on tests/coverage/pt_tempo_test.py:
```diff
@@ -123,72 +123,128 @@ def test_tempo_compute():
 def test_compute_dynamics_with_field():
```
I can't find the test file you refer to here

gefux replied:
That's because we've neglected to have a coverage test for anything in `contractions.py` so far. 
So, it would be great if you could create it.

piperfw commented on oqupy/tempo.py:
```diff
@@ -483,143 +489,245 @@ class TempoWithField(BaseTempo):
+        # Prepare the MeanFieldTempo backend
+        self._prepare_backend()
+
+    # These properties should move to TempoParameters when adaptive propagator
+    # construction has been added to Tempo. It doesn't make sense to allow
+    # these properties to be changed unless propagator_list is regenerated
```
Agreed, I am working on this now

piperfw commented:
Hi @gefux, thanks for all the suggestions.

I've made most changes suggested and implemented integration of 'regular' time-dependent system Liouvillians as in #41. I intend to check over this code more carefully this afternoon, but have pushed changes in case you wish to give feedback or corrections are this stage, in particular on:
- get_propagators is now a system method (removed from tempo.py, contractions.py)
- the default subdiv_limit parameter is not None, meaning by default a time-dependent system Liouvillian will be integrated rather than sampled (the behaviour prior to the update)
- the tests pt_tempo_test.py and tempo_test.py contain duplicate code for testing the parameters
- I rebased to reduce bring the commit messages title under 50 characters, except for the very first commit above (I wasn't sure if there would be implications in doing this)

piperfw commented:
This PR is ready to be merged now that
- the changes from @gefux/pr-mean-field have been merged
- all tests pass
- the documentation has been updated, including a new citation for the published mean-field paper
- integration of time dependent Liouvillian's has been implemented and checked (when subdiv_lim = None all computational results are as before, when this is not None there is a small correction due to integration vs sampling to construct the propagators)

A version bump 0.3.3 -> 0.4.0 will follow the PR merge
